first draft

maryamariyan opened the pull request:
cc: @safern @davidfowl

maryamariyan committed:
first draft

maryamariyan commented on ClassLibrary1/AugmentingGenerator.cs:
```diff
@@ -0,0 +1,77 @@
+            {
+                // Business logic to decide what we're interested in goes here
+                if (syntaxNode is ClassDeclarationSyntax cds &&
+                    cds.Identifier.ValueText == "CustomOptions")
```
@davidfowl 's comment https://github.com/maryamariyan/configuration-binding-source-generation/pull/1#discussion_r501299721:

```
We want to look for calls to Bind and Configure instead of looking for CustomOptions.
```

maryamariyan replied:
Makes sense. not sure yet what would replace with CustomOptions here.

maryamariyan replied:
Perhaps we could parse and look for a pattern like "Bind\<T\>" and retrieve the value T from it and for any such T we find autogenerate the code blob in this file? in this case CustomOptions. Just a thought.

maryamariyan replied:
Or if there was an attribute that could be set on classes like CustomOptions then we could loop through all and generate such code for them

davidfowl replied:
> Perhaps we could parse and look for a pattern like "Bind<T>" and retrieve the value T from it and for any such T we find autogenerate the code blob in this file? in this case CustomOptions. Just a thought.

This is what I think we should try.

> Or if there was an attribute that could be set on classes like CustomOptions then we could loop through all and generate such code for them

This works but you'd also need the attribute to have the ability to point to types that you don't own:

```
[ConfigurationBinder(typeof(CustomOptions))]
```

safern replied:
> Perhaps we could parse and look for a pattern like "Bind" and retrieve the value T from it and for any such T we find autogenerate the code blob in this file? in this case CustomOptions. Just a thought.

I also think we should pursue this. Also, something to consider, there are some non-generic `Bind` methods on `IConfiguration`, which takes the instance as a parameter.

So I think we should inspect `.Bind` calls and either use the `T` or the instance type if the call is to the non-generic method.

davidfowl replied:
@safern I have an issue that I'm almost done filing that says we should add a generic Bind\<T\> overload for the sake of this implementation 😄

maryamariyan committed:
Update gives ambiguous match

maryamariyan committed:
Update: working with simple cases

- TODO:
- [ ] cleanup SyntaxReceiver logic
- [ ] improve Bind logic
- [ ] Test for complex types

maryamariyan committed:
update gitignore
